contracts/src/token/erc1155/extensions/burnable.rs

```diff
@@ -0,0 +1,324 @@
+
+    /// The approval is cleared when the token is burned. Relies on the
+    /// `_burn_batch` mechanism.
+    ///
+    /// # Arguments
+    ///
+    /// * `account` - Accounts to burn tokens from.
+    /// * `values` - All amount to be burnt.
+    /// * `token_ids` - All token id to be burnt.
+    ///
+    /// # Errors
+    ///
+    /// If the caller is not account address and the account has not been
+    /// approved, then the error [`Error::MissingApprovalForAll`] is
+    /// returned.
+    /// If `from` is the `Address::ZERO`, then the error
+    /// [`Error::InvalidSender`] is returned.
+    /// If length of `ids` is not equal to length of `values`, then the
+    /// error [`Error::InvalidArrayLength`] is returned.
+    /// If any of the `values` is greater than the balance of the respective
+    /// token from `tokens` of the `from` account, then the error
+    /// [`Error::InsufficientBalance`] is returned.
+    ///
+    /// # Requirements:
+    ///
+    /// * `token_id` must exist.
+    /// * The caller or account must own `token_id` or be an approved operator.
+    fn burn_batch(
+        &mut self,
+        account: Address,
+        token_ids: Vec<U256>,
+        values: Vec<U256>,
+    ) -> Result<(), Self::Error>;
+}
+
+impl IErc1155Burnable for Erc1155 {
+    type Error = Error;
+
+    fn burn(
+        &mut self,
+        account: Address,
+        token_id: U256,
+        value: U256,
+    ) -> Result<(), Self::Error> {
+        let sender = msg::sender();
+        if account != sender && !self.is_approved_for_all(account, sender) {
+            return Err(Error::MissingApprovalForAll(
+                ERC1155MissingApprovalForAll {
+                    owner: account,
+                    operator: sender,
+                },
+            ));
+        }
+        self._burn(account, token_id, value)?;
+        Ok(())
+    }
+
+    fn burn_batch(
+        &mut self,
+        account: Address,
+        token_ids: Vec<U256>,
+        values: Vec<U256>,
+    ) -> Result<(), Self::Error> {
+        let sender = msg::sender();
+        if account != sender && !self.is_approved_for_all(account, sender) {
+            return Err(Error::MissingApprovalForAll(
+                ERC1155MissingApprovalForAll {
+                    owner: account,
+                    operator: sender,
+                },
+            ));
+        }
+        self._burn_batch(account, token_ids, values)?;
+        Ok(())
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
+    use alloy_primitives::{address, Address, U256};
+    use stylus_sdk::msg;
+
+    use super::IErc1155Burnable;
+    use crate::token::erc1155::{
+        ERC1155InvalidSender, ERC1155MissingApprovalForAll, Erc1155, Error,
+        IErc1155,
+    };
+
+    const BOB: Address = address!("F4EaCDAbEf3c8f1EdE91b6f2A6840bc2E4DD3526");
+
+    pub(crate) fn random_token_ids(size: usize) -> Vec<U256> {
+        (0..size).map(|_| U256::from(rand::random::<u32>())).collect()
```

Comment:
I would say it is better for testing to not use random values but create ids from sequence.